lib/services/devices/deviceService.js

```diff
@@ -199,6 +357,104 @@ function createInitialEntity(deviceData, newDevice, callback) {
     request(options, createInitialEntityHandler(deviceData, newDevice, callback));
 }
 
+/**
+ * Creates the initial entity representing the device in the Context Broker. This is important mainly to allow the
+ * rest of the updateContext operations to be performed using an UPDATE action instead of an APPEND one.
+ *
+ * @param {Object} deviceData       Object containing all the deviceData needed to send the registration.
+ * @param {Object} newDevice        Device object that will be stored in the database.
+ */
+function createInitialEntity(deviceData, newDevice, callback) {
+    if (config.checkNgsi2()) {
+        createInitialEntityNgsi2(deviceData, newDevice, callback);
+    } else {
+        createInitialEntityNgsi1(deviceData, newDevice, callback);
+    }
+}
+
+/**
+ * Updates the entity representing the device in the Context Broker using NGSIv2.
+ *
+ * @param {Object} deviceData       Object containing all the deviceData needed to send the registration.
+ * @param {Object} updatedDevice    Device object that will be stored in the database.
+ */
+function updateEntityNgsi2(deviceData, updatedDevice, callback) {
+    var options = {
+        url: config.getConfig().contextBroker.url + '/v2/entities/' + String(deviceData.name) + '/attrs',
+        method: 'PUT',
+        json: {
+        },
+        headers: {
+            'fiware-service': deviceData.service,
+            'fiware-servicepath': deviceData.subservice,
+            'fiware-correlator': (domain.active && domain.active.corr) || uuid.v4()
+        }
+    };
+
+    function jsonConcat(json1, json2) {
+        for (var key in json2) {
+            if (json2.hasOwnProperty(key)) {
+                json1[key] = json2[key];
+            }
+        }
+    }
+
+    function formatAttributes(originalVector, staticAtts) {
+        var attributeList = {};
+
+        if (originalVector && originalVector.length) {
+            for (var i = 0; i < originalVector.length; i++) {
+                attributeList[originalVector[i].name] = {
+                    type: originalVector[i].type,
+                    value: getInitialValueForType(originalVector[i].type)
+                };
+
+                if (staticAtts) {
+                    attributeList[originalVector[i].name].value = originalVector[i].value;
+                } else {
+                    attributeList[originalVector[i].name].value = getInitialValueForType(originalVector[i].type);
+                }
+            }
+        }
+
+        return attributeList;
+    }
+
+    function formatCommands(originalVector) {
```

Comment:
I see several functions with the same name in different parts, in updateEntityNgsi2() and createInitialEntityNgsi2():

* formatCommands()
* formatAttributes()
* jsonCat()

Are they the same? Could be refactored into a common place to avoid duplication?

Reply:
Fixed with 8ce26bd8f52505a4fd12bd42bded8630613bd7a1